### ISSUE TITLE
Not OK: perl v5.7.2 +DEVEL13165 on armv4l-linux-64int 2.4.5-rmk6    (UNINSTALLED)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4599]
Migrated from [rt.perl.org#7925](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7925) (status was 'resolved')

Searchable as RT7925$

**Comment by p5pRT** [comment]
### From @nwc10



lib/Net/Ping/t/200\_ping\_tcp\.t starts&#8203;:

&nbsp;    \# Remote network test using tcp protocol\.
&nbsp;    \#
&nbsp;    \# NOTE&#8203;:
&nbsp;    \#   Network connectivity will be required for all tests to pass\.
&nbsp;    \#   Firewalls may also cause some tests to fail\, so test it
&nbsp;    \#   on a clear network\.

This failed&#8203;:

&nbsp;    \# Test on the default port
&nbsp;    ok $p \-> ping\("localhost"\);

Why does it assume that I'm running tcp echo service? Shutting down echo\,
discard and chargen is one of the first things I do\.

&nbsp;    \# Hopefully this is not a routeable host
&nbsp;    ok \!$p \-> ping\("10\.12\.14\.16"\);

Not great assuming things about my private network\. However it did pass as
the only 10's around here are 10\.153\.206\.39\, 36 and 2

&nbsp;    \# Test a few remote servers
&nbsp;    \# Hopefully they are up when the tests are run\.

&nbsp;    ok $p \-> ping\("www\.geocities\.com"\);
&nbsp;    ok $p \-> ping\("ftp\.geocities\.com"\);

&nbsp;    ok $p \-> ping\("www\.freeservers\.com"\);
&nbsp;    ok $p \-> ping\("ftp\.freeservers\.com"\);

&nbsp;    ok $p \-> ping\("yahoo\.com"\);
&nbsp;    ok $p \-> ping\("www\.yahoo\.com"\);
&nbsp;    ok $p \-> ping\("www\.about\.com"\);

All these failed\. My PPP link is down\. Mine doesn't autodial\.
Is it a great idea make a regression test that could trigger autodials?

Nicholas Clark
 

<details><summary>Perl Info</summary>

```

Flags:
    category=install
    severity=none

Site configuration information for perl v5.7.2:

Configured by nick at Wed Nov 21 20:16:31 GMT 2001.

Summary of my perl5 (revision 5.0 version 7 subversion 2 patch 13178) configuration:
  Platform:
    osname=linux, osvers=2.4.5-rmk6, archname=armv4l-linux-64int
    uname='linux bagpuss.unfortu.net 2.4.5-rmk6 #4 tue jun 19 23:26:57 bst 2001 armv4l unknown '
    config_args='-Doptimize=-O1 -Dusedevel -Uinstallusrbinperl -Dcf_email=nick@talking.bollo.cx -Dperladmin=nick@talking.bollo.cx -Dinc_version_list=  -Dinc_version_list_init=0 -Duse64bitint -de'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=define
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O1',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.4 20010319 (Debian prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL13165


@INC for perl v5.7.2:
    lib
    /usr/local/lib/perl5/5.7.2/armv4l-linux-64int
    /usr/local/lib/perl5/5.7.2
    /usr/local/lib/perl5/site_perl/5.7.2/armv4l-linux-64int
    /usr/local/lib/perl5/site_perl/5.7.2
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.7.2:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=en_GB.ISO-8859-1
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/sbin:/usr/sbin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/sh

```
</details>

**Comment by p5pRT** [comment]
### From @jhi

Yes\.  In general I think we have no choice but to ship Net&#8203;::Ping
in the core distribution without its tests \(or most of them\)\.  There
simply is no way we can require any kind of networking setup \(\*\)\,
including requiring any network\.

\(\*\) Before the testing posse is on my case :\-\) let me preempt them by
saying that fake networks don't really help here\.  You'll be testing
s fake network\.  That proves that the module worked on a fake network\.
That's great\.  Like drinking decaffeinated\.

> this failed&#8203;:
> 
>     \# Test on the default port
>     ok $p \-> ping\("localhost"\);
> 
> Why does it assume that I'm running tcp echo service? Shutting down echo\,
> discard and chargen is one of the first things I do\.
> 
>     \# Hopefully this is not a routeable host
>     ok \!$p \-> ping\("10\.12\.14\.16"\);
> 
> Not great assuming things about my private network\. However it did pass as
> the only 10's around here are 10\.153\.206\.39\, 36 and 2
> 
>     \# Test a few remote servers
>     \# Hopefully they are up when the tests are run\.
> 
>     ok $p \-> ping\("www\.geocities\.com"\);
>     ok $p \-> ping\("ftp\.geocities\.com"\);
> 
>     ok $p \-> ping\("www\.freeservers\.com"\);
>     ok $p \-> ping\("ftp\.freeservers\.com"\);
> 
>     ok $p \-> ping\("yahoo\.com"\);
>     ok $p \-> ping\("www\.yahoo\.com"\);
>     ok $p \-> ping\("www\.about\.com"\);
> 
> All these failed\. My PPP link is down\. Mine doesn't autodial\.
> Is it a great idea make a regression test that could trigger autodials?
> 
> Nicholas Clark
>  
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=install

\-\- 
$jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
&nbsp;        \# There is this special biologist word we use for 'stable'\.
&nbsp;        \# It is 'dead'\. \-\- Jack Cohen

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Can we prompt the user to ask if they have a network?

&nbsp;    print "Do you have a clear network I can test Net&#8203;::Ping on? \(will
assume no in 5 seconds\) ";
&nbsp;    $SIG\{ALRM\}=\\&bail;
&nbsp;    alarm\(5\);        \#although alarm\(\) doesn't work everywhere\.\.\.
&nbsp;    my $response=\<STDIN>;
&nbsp;    bail\(\) if\($response \!~ /y/i\);

\-\-Brent Dax
brentdax@&#8203;cpan\.org
Configure pumpking for Perl 6

"Nothing important happened today\."
&nbsp;    \-\-George III of England's diary entry for 4\-Jul\-1776

**Comment by p5pRT** [comment]
### From @jhi

No\.  Prompting bad\.

Besides\, what the heck is a "clear network"?  As you can see from
Nick's travails a network is a sum of many variables&#8203;: many servers\,
firewalls\, proxies\.  We can't know what those variables are\.  We
shouldn't guess\.  Let's move on and find something we can solve\.

>     print "Do you have a clear network I can test Net&#8203;::Ping on? \(will
> assume no in 5 seconds\) ";
>     $SIG\{ALRM\}=\\&bail;
>     alarm\(5\);        \#although alarm\(\) doesn't work everywhere\.\.\.
>     my $response=\<STDIN>;
>     bail\(\) if\($response \!~ /y/i\);
> 
> \-\-Brent Dax
> brentdax@&#8203;cpan\.org
> Configure pumpking for Perl 6
> 
> "Nothing important happened today\."
>     \-\-George III of England's diary entry for 4\-Jul\-1776

\-\- 
$jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
&nbsp;        \# There is this special biologist word we use for 'stable'\.
&nbsp;        \# It is 'dead'\. \-\- Jack Cohen

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

> Jarkko Hietaniemi&#8203;:

> \# \(\*\) Before the testing posse is on my case :\-\) let me preempt them by 
> \# saying that fake networks don't really help here\.  You'll be testing 
> \# s fake network\.  That proves that the module worked on a fake network\.
> \# That's great\.  Like drinking decaffeinated\.

> Can we prompt the user to ask if they have a network?

Yes\, but that would make the test suite interactive \(yikes\) or reliant on an
unportable alarm\(\) call\.

It's still a kind of chicken\-and\-egg problem\.  The only way to test this
reliably is to fake up some socket stuff underneath\. Otherwise\, it's extremely
difficult to write portable code without the kind of assumptions that Nicholas
Clark found\.

Like Jarkko points out\, the best test I could write would only be a
semi\-guarantee that\, if the user has a network that works like the test expects\,
the module should work there\.  Depending on how well socket\(\) and recv\(\) and
such are tested\, that may or may not be a correct assumption\.

It's hard to see a good way to solve this without keeping track of more
capabilities\.  That could be a lot of rejiggering\.

Thinking about this one evening gave me a headache\.

Back to DB\.t\, 
\-\- c

**Comment by p5pRT** [comment]
### From @jhi

\-\- 
$jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
&nbsp;        \# There is this special biologist word we use for 'stable'\.
&nbsp;        \# It is 'dead'\. \-\- Jack Cohen

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Moin\,

Jarkko wrote&#8203;:
>On Wed\, Nov 21\, 2001 at 11&#8203;:55&#8203;:40PM \+0000\, Nicholas Clark wrote&#8203;:
>> lib/Net/Ping/t/200\_ping\_tcp\.t starts&#8203;:
>> 
>>     \# Remote network test using tcp protocol\.
>>     \#
>>     \# NOTE&#8203;:
>>     \#   Network connectivity will be required for all tests to pass\.
>>     \#   Firewalls may also cause some tests to fail\, so test it
>>     \#   on a clear network\.
>
>Yes\.  In general I think we have no choice but to ship Net&#8203;::Ping
>Min the core distribution without its tests \(or most of them\)\.  There
>simply is no way we can require any kind of networking setup \(\*\)\,
>including requiring any network\.
>
>\(\*\) Before the testing posse is on my case :\-\) let me preempt them by
>saying that fake networks don't really help here\.  You'll be testing
>s fake network\.  That proves that the module worked on a fake network\.
>That's great\.  Like drinking decaffeinated\.

But if you'r designing a coffee drinking module\, it is stil better to
ensure that it can at least hold the cup \(even if it is a plastic one\)\,
drink something\, and not spill it\. Even if it is not real coffee\.\[0\]

Still better than an entirely untested module\, which breaks in the brandnew
5\.8\.0 because 10 seconds before release a pod patch introduces a syntax
error and nobody notices until it is too late\. 

If you get my point ;\)

Cheers\,

Tels

\[0\] Even the Math&#8203;::Bigint tests are "fake" in the sense\, that they assert
that 5\+5 is 10\, but they don't assure that it will really work with real
numbers\, like 2 \*\* 654321\+1\. They just \(try to\) build confidence that if it
works with 2\+2=4\, it will work for PI to 100\.000 digits\. Same goes for
network modules\, if it works with the fake thing\, thats better than if it
doesn't work at all and we have hopes that it will pass the real thing\.\[1\]

\[1\] That means\, of course\, that the fake thing doesn't really emulate the
real thing 100%\. In this case\, the tests \_are\_ for real\, because neither
the test nor the usage could differ between a fake and a real thing\. Of
course\, in praxis emulations are mostly\.\.uh\.\.incomplete\. But better than
nothing\, but I repeat myself\, so I stop now ;\)





\- \-\- 
perl \-MMath&#8203;::String \-e 'print \\
Math&#8203;::String\->from\_number\("215960156869840440586892398248"\)\,"\\n"'

 http&#8203;://bloodgate\.com/perl       My current Perl projects
 PGP key available on http&#8203;://bloodgate\.com/tels\.asc or via email\.


\-\-\-\-\-BEGIN PGP SIGNATURE\-\-\-\-\-
Version&#8203;: 2\.6\.3i
Charset&#8203;: latin1

iQEVAwUBO/ylZHcLPEOTuEwVAQEZhQf9E9Msk\+O4yU5s/1UlUE2GBxpRRsizoI/z
8zWjXbRYkiLrKE3ZhtQi8VAwSSgcXkThTKtSu4HGETW63T4H0Dn\+rc2wYoRQc47D
042l0xHcgHRuRkQISCWqosyA869Jsxh2OBkod/naKs5629j0r\+gUJ9xJMe0XP/8R
AO8lZhhIBRMUwqs1vzqoEl3734OLeIMBCxklmPXnkthxblsbHYgrcepfyMwhVeOR
o5mgZUbmX0Zc7fpxzEueeQSE1gGz2cwT0ShYzHJn4jDgOYnORx903SDF6F6kS\+lf
vtkg3qX3D3Xh\+qwvx0G2bHRvtaLQCmWaOtPEt8I5xKwCvZHG3n\+lhg==
=mOXN
\-\-\-\-\-END PGP SIGNATURE\-\-\-\-\-

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Mine does but ping's default timeout fails before it comes up\.

>Is it a great idea make a regression test that could trigger autodials?

FWIW \(probably due to my dubious DNS scheme\) hostname test already does
for me\.

>
>Nicholas Clark
>
>\[Please do not change anything below this line\]
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\-\-\-
>Flags&#8203;:
>    category=install
>    severity=none
>\-\-\-
>Site configuration information for perl v5\.7\.2&#8203;:
>
>Configured by nick at Wed Nov 21 20&#8203;:16&#8203;:31 GMT 2001\.
>
>Summary of my perl5 \(revision 5\.0 version 7 subversion 2 patch 13178\) configuration&#8203;:
>  Platform&#8203;:
>    osname=linux\, osvers=2\.4\.5\-rmk6\, archname=armv4l\-linux\-64int
>    uname='linux bagpuss\.unfortu\.net 2\.4\.5\-rmk6 \#4 tue jun 19 23&#8203;:26&#8203;:57 bst 2001 armv4l unknown '
>    config\_args='\-Doptimize=\-O1 \-Dusedevel \-Uinstallusrbinperl \-Dcf\_email=nick@&#8203;talking\.bollo\.cx \-Dperladmin=nick@&#8203;talking\.bollo\.cx \-Dinc\_version\_list=  \-Dinc\_version\_list\_init=0 \-Duse64bitint \-de'
>    hint=recommended\, useposix=true\, d\_sigaction=define
>    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
>    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>    use64bitint=define use64bitall=undef uselongdouble=undef
>    usemymalloc=n\, bincompat5005=define
>  Compiler&#8203;:
>    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
>    optimize='\-O1'\,
>    cppflags='\-fno\-strict\-aliasing \-I/usr/local/include'
>    ccversion=''\, gccversion='2\.95\.4 20010319 \(Debian prerelease\)'\, gccosandvers=''
>    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=12345678
>    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=8
>    ivtype='long long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
>    alignbytes=4\, prototype=define
>  Linker and Libraries&#8203;:
>    ld='cc'\, ldflags =' \-L/usr/local/lib'
>    libpth=/usr/local/lib /lib /usr/lib
>    libs=\-lnsl \-lgdbm \-ldbm \-ldb \-ldl \-lm \-lc \-lcrypt \-lutil
>    perllibs=\-lnsl \-ldl \-lm \-lc \-lcrypt \-lutil
>    libc=/lib/libc\-2\.2\.4\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
>  Dynamic Linking&#8203;:
>    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
>    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'
>
>Locally applied patches&#8203;:
>    DEVEL13165
>
>\-\-\-
>@&#8203;INC for perl v5\.7\.2&#8203;:
>    lib
>    /usr/local/lib/perl5/5\.7\.2/armv4l\-linux\-64int
>    /usr/local/lib/perl5/5\.7\.2
>    /usr/local/lib/perl5/site\_perl/5\.7\.2/armv4l\-linux\-64int
>    /usr/local/lib/perl5/site\_perl/5\.7\.2
>    /usr/local/lib/perl5/site\_perl
>    \.
>
>\-\-\-
>Environment for perl v5\.7\.2&#8203;:
>    HOME=/home/nick
>    LANG \(unset\)
>    LANGUAGE \(unset\)
>    LC\_CTYPE=en\_GB\.ISO\-8859\-1
>    LD\_LIBRARY\_PATH \(unset\)
>    LOGDIR \(unset\)
>    PATH=/home/nick/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/bin/X11&#8203;:/usr/games&#8203;:/sbin&#8203;:/usr/sbin&#8203;:/usr/local/sbin
>    PERL\_BADLANG \(unset\)
>    SHELL=/bin/sh
\-\-
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/
